File: 04-Queues/java/src/Main.java
import java.util.ArrayDeque;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.Stack;

public class Main {
    public static void reverse(Queue<Integer> queue) {
        Stack<Integer> stack = new Stack<>();
        while (!queue.isEmpty())
            stack.push(queue.remove());
        while (!stack.isEmpty())
            queue.add(stack.pop());
    }

    public static void main(String[] args) {
//        built-in ArrayDeque
        System.out.println("built-in ArrayDeque");

        Queue<Integer> queue = new ArrayDeque<>();
        queue.add(10);
        queue.add(20);
        queue.add(30);

        var front = queue.remove();
        System.out.println(front);
        System.out.println(queue);

        System.out.println("--------");

//        reverse
        System.out.println("reverse");

        reverse(queue);
        System.out.println(queue);

        System.out.println("--------");

//        custom ArrayQueue
        System.out.println("custom ArrayQueue");

        ArrayQueue arrayQueue = new ArrayQueue(5);
        arrayQueue.enqueue(5);
        arrayQueue.enqueue(6);
        arrayQueue.enqueue(7);
        arrayQueue.enqueue(8);
        arrayQueue.enqueue(9);
        System.out.println(arrayQueue);
        arrayQueue.dequeue();
        System.out.println(arrayQueue);
        arrayQueue.enqueue(10);
        System.out.println(arrayQueue);

        System.out.println("--------");

//        QueueWithTwoStacks
        System.out.println("QueueWithTwoStacks");

        QueueWithTwoStacks queueWithTwoStacks = new QueueWithTwoStacks();
        queueWithTwoStacks.enqueue(10);
        queueWithTwoStacks.enqueue(20);
        queueWithTwoStacks.enqueue(30);
        System.out.println(queueWithTwoStacks.dequeue());
        System.out.println(queueWithTwoStacks.dequeue());
        System.out.println(queueWithTwoStacks.peek());

        System.out.println("--------");

//        built-in priority queue
        System.out.println("built-in priority");
        PriorityQueue<Integer> priorityQueue = new PriorityQueue<>();
        priorityQueue.add(5);
        priorityQueue.add(1);
        priorityQueue.add(3);
        priorityQueue.add(2);
        while (!priorityQueue.isEmpty())
            System.out.println(priorityQueue.remove());

        System.out.println("--------");

//        custom priority queue
        System.out.println("custom priority queue");
        MyPriorityQueue myPriorityQueue = new MyPriorityQueue(5);
        myPriorityQueue.add(3);
        myPriorityQueue.add(7);
        myPriorityQueue.add(5);
        myPriorityQueue.add(1);
        myPriorityQueue.add(4);
        System.out.println(myPriorityQueue);

        while (!myPriorityQueue.isEmpty())
            System.out.println(myPriorityQueue.remove());

        System.out.println("--------");

//        QueueReverser
        System.out.println("QueueReverser");

        QueueReverser queueReverser = new QueueReverser();
        Queue<Integer> testQueue = new ArrayDeque<>();

        testQueue.add(10);
        testQueue.add(20);
        testQueue.add(30);
        testQueue.add(40);
        testQueue.add(50);
        System.out.println(testQueue);

        queueReverser.reverse(testQueue, 3);
        System.out.println(testQueue);

        System.out.println("--------");

//        LinkedListQueue
        System.out.println("LinkedListQueue");

        LinkedListQueue linkedListQueue = new LinkedListQueue();
        linkedListQueue.enqueue(10);
        linkedListQueue.enqueue(20);
        linkedListQueue.enqueue(30);
        linkedListQueue.enqueue(40);
        System.out.println(linkedListQueue.peek());
        System.out.println(linkedListQueue.dequeue());
        System.out.println(linkedListQueue.dequeue());
        System.out.println(linkedListQueue.dequeue());
        System.out.println(linkedListQueue.dequeue());
        System.out.println(linkedListQueue.isEmpty());

        System.out.println("--------");

//        StackWithTwoQueues
        System.out.println("StackWithTwoQueues");

        StackWithTwoQueues stackWithTwoQueues = new StackWithTwoQueues();
        stackWithTwoQueues.push(10);
        stackWithTwoQueues.push(30);
        stackWithTwoQueues.push(50);

        System.out.println(stackWithTwoQueues.pop());
        System.out.println(stackWithTwoQueues.peek());

        System.out.println(stackWithTwoQueues);

        System.out.println("--------");


    }

}

File: 04-Queues/cpp/main.cpp
#include <iostream>
#include <queue>
#include <stack>
#include <string>
#include "ArrayQueue.h"
#include "QueueWithTwoStacks.h"
#include "MyPriorityQueue.h"
#include "QueueReverser.h"
#include "LinkedListQueue.h"
#include "StackWithTwoQueues.h"

using namespace std;

void myReverse(queue<int> &);

int main() {
    // built-in queue

    cout << "built-in queue" << endl;
    queue<int> q;
    q.push(10);
    q.push(20);
    q.push(30);

    cout << "--------" << endl;

//    while (!q.empty()) {
//        cout << q.front() << endl;
//        q.pop();
//    }

    // myReverse
    cout << "myReverse" << endl;

    myReverse(q);
    cout << "After myReverse..." << endl;
    while (!q.empty()) {
        cout << q.front() << endl;
        q.pop();
    }
    cout << "--------" << endl;

    // Custom ArrayQueue

    cout << "Custom ArrayQueue" << endl;

    ArrayQueue arrayQueue(5);
    arrayQueue.enqueue(5);
    arrayQueue.enqueue(6);
    arrayQueue.enqueue(7);
    arrayQueue.enqueue(8);
    arrayQueue.enqueue(9);

    cout << arrayQueue.toString() << endl;
    cout << arrayQueue.dequeue() << endl;
    cout << arrayQueue.toString() << endl;
    arrayQueue.enqueue(10);
    cout << arrayQueue.toString() << endl;

    cout << "--------" << endl;

    // QueueWithTwoStacks

    cout << "QueueWithTwoStacks" << endl;

    QueueWithTwoStacks queueWithTwoStacks;
    queueWithTwoStacks.enqueue(10);
    queueWithTwoStacks.enqueue(20);
    queueWithTwoStacks.enqueue(30);

    cout << queueWithTwoStacks.peek() << endl;
    cout << queueWithTwoStacks.dequeue() << endl;
    cout << queueWithTwoStacks.dequeue() << endl;
    cout << queueWithTwoStacks.dequeue() << endl;
    cout << "--------" << endl;

    // built-in priority queue

    cout << "built-in priority queue" << endl;
    priority_queue<int, vector<int>, greater<int>> priorityQueue;
    priorityQueue.push(7);
    priorityQueue.push(5);
    priorityQueue.push(15);
    priorityQueue.push(25);
    while (!priorityQueue.empty()) {
        cout << priorityQueue.top() << endl;
        priorityQueue.pop();
    };

    cout << "--------" << endl;

    // custom priority queue
    cout << "custom priority queue" << endl;

    MyPriorityQueue myPriorityQueue(5);
    myPriorityQueue.add(3);
    myPriorityQueue.add(7);
    myPriorityQueue.add(5);
    myPriorityQueue.add(1);
    myPriorityQueue.add(4);

    cout << myPriorityQueue.toString() << endl;

    while (!myPriorityQueue.isEmpty())
        cout << myPriorityQueue.remove() << endl;

    cout << "--------" << endl;

    // QueueReverser
    cout << "QueueReverser" << endl;

    QueueReverser queueReverser;
    queue<int> testQueue;

    testQueue.push(10);
    testQueue.push(20);
    testQueue.push(30);
    testQueue.push(40);
    testQueue.push(50);

    queueReverser.reverse(testQueue, 3);

    while (!testQueue.empty()) {
        cout << testQueue.front() << endl;
        testQueue.pop();
    }

    cout << "--------" << endl;

    // LinkedListQueue
    cout << "LinkedListQueue" << endl;

    LinkedListQueue linkedListQueue;
    linkedListQueue.enqueue(10);
    linkedListQueue.enqueue(30);
    linkedListQueue.enqueue(50);
    cout << linkedListQueue.toString() << endl;
    cout << linkedListQueue.dequeue() << endl;
    cout << linkedListQueue.dequeue() << endl;
    cout << linkedListQueue.peek() << endl;

    cout << "--------" << endl;

    // StackWithTwoQueues
    cout << "StackWithTwoQueues" << endl;

    StackWithTwoQueues stackWithTwoQueues;

    stackWithTwoQueues.push(10);
    stackWithTwoQueues.push(30);
    stackWithTwoQueues.push(50);
    cout << stackWithTwoQueues.pop() << endl;
    cout << stackWithTwoQueues.peek() << endl;
    cout << stackWithTwoQueues.toString() << endl;

    cout << "--------" << endl;


    return 0;
}

void myReverse(queue<int> &queue) {
    stack<int> st;
    while (!queue.empty()) {
        st.push(queue.front());
        queue.pop();
    }
    while (!st.empty()) {
        queue.push(st.top());
        st.pop();
    }

}